### DISCUSSION TITLE
ESP32 + Relay + integrated 230V AC Power Supply

### DISCUSSION THREAD

**Discussion by Noschvie** [arendst/Tasmota#21056] [Q&A] [answered]
Hello

found this [ESP32 + relay + integrated 230V AC](https://f1atb.fr/esp32-relay-integrated-230v-ac-power-supply-sensors/) device and want to give it a try...

Configuration:
![grafik](https://github.com/arendst/Tasmota/assets/48217146/44b22902-47f7-4a05-912e-473c41a6a7aa)
**But:** Restarting the board results in switching the relay.
It seems to be `ESP32-D0WD-V3 v3.1` related, but why? Are there restrictions on the use of the GPIO 16 ?

Thanks!

**Comment by valdus2** [comment]
Did you try poweron state?

**Comment by Noschvie** [comment]
Yes, I did.  `PowerOnState 1` dosn't work. But using `PowerOnState 3` the relay should not "flip - flop".

```
00:00:00.003 HDW: ESP32-D0WD-V3 v3.1 
00:00:00.064 UFS: FlashFS mounted with 304 kB free
00:00:00.097 CFG: Loaded from File, Count 67
00:00:00.107 QPC: Count 1
00:00:00.220 BRY: Berry initialized, RAM used 4446 bytes
00:00:00.244 Project tasmota - Tasmota DevKit One Version 13.4.0(tasmota32)-2_0_14(2024-02-14T16:17:40)
00:00:01.033 WIF: Connecting to AP1 neunzehn.iot Channel 6 BSSId 40:E3:D6:4B:99:C0 in mode 11n as tasmota-824244-0580...
00:00:03.672 WIF: Connected
00:00:03.881 HTP: Web server active on tasmota-824244-0580 with IP address 192.168.7.113
20:52:09.882 RSL: INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"13.4.0(tasmota32)","FallbackTopic":"cmnd/DVES_824244_fb/","GroupTopic":"cmnd/tasmotas/"}}
20:52:09.897 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-824244-0580","IPAddress":"192.168.7.113","IP6Global":"","IP6Local":"fe80::e665:b8ff:fe82:4244%st1"}}
20:52:09.910 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":21}}
20:52:09.913 RSL: RESULT = {"POWER":"OFF"}
20:52:09.924 RSL: POWER = OFF
20:52:12.922 QPC: Reset
20:52:13.920 RSL: STATE = {"Time":"2024-03-28T20:52:13","Uptime":"0T00:00:08","UptimeSec":8,"Heap":154,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"POWER":"OFF","Wifi":{"AP":1,"SSId":"neunzehn.iot","BSSId":"40:E3:D6:4B:99:C0","Channel":6,"Mode":"11n","RSSI":30,"Signal":-85,"LinkCount":1,"Downtime":"0T00:00:03"}}
20:52:44.161 CMD: poweronstate 
20:52:44.164 RSL: RESULT = {"PowerOnState":1}

```

**Reply by sfromis** [reply, marked as answer]
Using gpio 16 for a relay is a bad plan, as gpio 16/17 are commonly used for PSRAM, and therefore probed during boot.

**Reply by Noschvie** [reply]
I feared that... Thanks again.

**Reply by rebel7580** [reply]
You might be able to rewire which IO the relay connects to. From the board photos on AliExpress, it looks like IO16, which also goes to a header, has a "through hole" connection to R16, which in turn is connected to the base of the Q1 transistor, which drives the relay.
If you (carefully) remove R16, that would remove IO16 from the relay. Then you could connect to a different IO, through an external resistor, to where R16 used to be. Just an idea.

An alternative method might be to drill out the through hole, which would break the connection between IO16 and R16, then simply solder-tack a wire from R16 to another IO. Wouldn't have to remove R16.

SWAGing this of course. YMMV.

**Reply by Noschvie** [reply]
Thanks for your replay. Will try to remove R16, but I need a magnifying glass and a tweezers...

**Reply by Noschvie** [reply]
![image](https://github.com/arendst/Tasmota/assets/48217146/fb621083-a8f2-4111-9a36-71c4e50f55a4)

On the backside there seems to be the line to GPIO16. Maybe this should be easy to cut.

**Reply by sfromis** [reply]
Looks like a custom build adding `-DFRAMEWORK_ARDUINO_ITEAD` to `build_flags` changes what pins are polled for PSRAM detection. However, this is about gpio 17, not about gpio 16 which is not movable.

**Reply by Noschvie** [reply]
Changed `platformio_override.ini` and added the build_flags to the tasmota32_base section. But no change for GPIO 16.

```
[env:tasmota32_base]
build_flags             = ${common.build_flags}
                          -DFRAMEWORK_ARDUINO_ITEAD
```

**Reply by rebel7580** [reply]
@Noschvie  - Yes, that's the GPIO16 line that goes to R16 (check with a meter to make sure it is indeed R16 and not the one right next to it), and that might be the easiest approach. Less likely to cause damage to adjacent areas, and more easily reversible if needed.

**Comment by wjohn007** [comment]
@Noschvie
Please take a look to [blakadder](https://templates.blakadder.com/ESP32_Relay_AC_X1.html)

`Flash with tasmota32-nspanel.bin, otherwise the relay on the board clicks on reboot because it’s using a pin reserved for PSRAM.`

**Reply by Noschvie** [reply]
@wjohn007 thanks for this link!

in case of the relay state is "Off" and the board is restarting the boards doesn't click on reboot. But if the relay in "On" the click on reboot isn't suppressed.
